IntegrationTests/UITests/UITest.cs

```diff
@@ -27,32 +30,47 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
-            Func<T, Task> runAsync)
+            Action showForm,
+            Func<T, Task> runTestAsync)
             where T : Form
         {
-            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            // If form.Dipose was called by the async test task, closing the window would be done in a strange order.
+            T form = null;
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
-                try
-                {
-                    await runAsync(dialog);
-                }
-                finally
+                // Start runTestAsync before calling showForm.
+                // The latter might block until the form is closed, especially if using Application.Run(form).
+
+                // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+                // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+                // complete.
+                var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    dialog.Close();
-                }
-            });
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    await WaitForIdleAsync();
+                    form = Application.OpenForms.OfType<T>().Single();
+                    try
+                    {
+                        Assert.True(form.Visible, $"{form.GetType().FullName} {form.Name} should be visible");
```

Comment:
💭 This seems like an unnecessary assertion. Is there a particular problem if the form is not visible?

Reply:
Yes, this might be not necessary assertion. It can be removed when the tests run stable.
Though it is fulfilled for all GE tests.
I am still afraid of a race condition in `RunForm` / `WaitForIdleAsync`, e.g. in case `showForm` does not directly open a Form but runs (asynchronous) commands before.

Reply:
`showForm` is required to open the form synchronously.

Reply:
> `showForm` is required to open the form synchronously.

That's clear. But before it opens the Form, it could do some preparation before, e.g. using `RunAsync`.
Is it really ensured that the event `Application.Idle` is not emitted before the main thread starts a message loop by opening a Form?